### ISSUE TITLE
Legacy of Kain - Unexpected reboot may occur when loading certain movies

### ISSUE THREAD

**Issue by kihato** [flyinghead/flycast#1602]
Platform / OS / Hardware: 
Windows11 / i7-10700F / GTX 1660 SUPER

Github hash: 6061d40

https://github.com/user-attachments/assets/879f29b6-a583-44b9-8490-62757d5564a9

[Legacy of Kain - Soul Reaver v1.000 (1999)(EIDOS)(NTSC)(US)[!].state.zip](https://github.com/user-attachments/files/16675874/Legacy.of.Kain.-.Soul.Reaver.v1.000.1999.EIDOS.NTSC.US.state.zip)

As you progress through the corridors, a total of six forced cutscenes will play.
It's not clear which cutscene will trigger a reboot, and sometimes there won't be a reboot at all.

The early January 2024 version seems stable,
but we have confirmed that reboots occur without fail after that.

**Comment by bslenul** [comment]
I suck at debugging so sorry in advance if all that follows doesn't help... 😳 

Anyway, after reloading the state a few times I was able to reproduce and saw that in logs: `01:56:040 hw/holly/sb.cpp:220 N[SH4]: SB/HOLLY: System reset requested` so I started GDB and put a breakpoint at this line and here's what I got:
```
Thread 45 "Flycast-emu" hit Breakpoint 1, sb_write_SB_SFRES (addr=2690607248, data=30225) at G:/msys64/home/B-S/flycast/core/hw/holly/sb.cpp:220
220			NOTICE_LOG(SH4, "SB/HOLLY: System reset requested");
#0  sb_write_SB_SFRES (addr=2690607248, data=30225) at G:/msys64/home/B-S/flycast/core/hw/holly/sb.cpp:220
No locals.
#1  sb_write_SB_SFRES (addr=2690607248, data=30225) at G:/msys64/home/B-S/flycast/core/hw/holly/sb.cpp:216
No locals.
#2  0x00007ff6e00acfc0 in ?? ()
No symbol table info available.
#3  0x00007ffabb735880 in RegisterWaitForSingleObject () from C:\Windows\System32\kernel32.dll
No symbol table info available.
#4  0x0000000000000000 in ?? ()
No symbol table info available.
[Switching to thread 1 (Thread 3104.0x31e4)]
#0  0x00007ffabd38e0b4 in ntdll!ZwWaitForMultipleObjects () from C:\Windows\SYSTEM32\ntdll.dll
#0  0x00007ffabd38e0b4 in ntdll!ZwWaitForMultipleObjects () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#1  0x00007ffabb084030 in WaitForMultipleObjectsEx () from C:\Windows\System32\KernelBase.dll
No symbol table info available.
#2  0x00007ffabb083f2e in WaitForMultipleObjects () from C:\Windows\System32\KernelBase.dll
No symbol table info available.
#3  0x00007ff6e120c688 in _pthread_wait_for_multiple_objects ()
No symbol table info available.
#4  0x00007ff6e120b5c0 in do_sema_b_wait_intern ()
No symbol table info available.
#5  0x00007ff6e120b896 in do_sema_b_wait ()
No symbol table info available.
#6  0x00007ff6e120c2b8 in pthread_cond_timedwait_impl ()
No symbol table info available.
#7  0x00007ff6e10e1176 in __gthread_cond_timedwait (__abs_timeout=0x5feea0, __mutex=0x7ff6e2e0bd08 <pvrQueue+8>, __cond=0x7ff6e2e0bd10 <pvrQueue+16>) at G:/msys64/mingw64/include/c++/13.2.0/x86_64-w64-mingw32/bits/gthr-default.h:872
No locals.
#8  std::__condvar::wait_until (__abs_time=..., __m=..., this=0x7ff6e2e0bd10 <pvrQueue+16>) at G:/msys64/mingw64/include/c++/13.2.0/bits/std_mutex.h:178
No locals.
#9  std::condition_variable::__wait_until_impl<std::chrono::duration<long long, std::ratio<1ll, 1000000000ll> > > (__lock=<synthetic pointer>..., __atime=Python Exception <class 'gdb.error'>: value has been optimized out
<synthetic pointer>, this=0x7ff6e2e0bd10 <pvrQueue+16>) at G:/msys64/mingw64/include/c++/13.2.0/condition_variable:224
        __s = <optimized out>
        __ns = <optimized out>
        __ts = {tv_sec = 1724588038, tv_nsec = 735102400}
        __s = <optimized out>
        __ns = <optimized out>
        __ts = <optimized out>
#10 std::condition_variable::wait_until<std::chrono::_V2::steady_clock, std::chrono::duration<long long, std::ratio<1ll, 1000000000ll> > > (__atime=..., __lock=<synthetic pointer>..., this=0x7ff6e2e0bd10 <pvrQueue+16>) at G:/msys64/mingw64/include/c++/13.2.0/condition_variable:137
        __c_entry = std::chrono::_V2::steady_clock time_point = { 328928275719500ns }
        __s_entry = <optimized out>
        __delta = <optimized out>
        __s_atime = <optimized out>
#11 std::condition_variable::wait_for<long long, std::ratio<1ll, 1000ll> > (__rtime=..., __lock=<synthetic pointer>..., this=0x7ff6e2e0bd10 <pvrQueue+16>) at G:/msys64/mingw64/include/c++/13.2.0/condition_variable:165
No locals.
#12 cResetEvent::Wait (this=this@entry=0x7ff6e2e0bd08 <pvrQueue+8>, msec=msec@entry=20) at G:/msys64/home/B-S/flycast/core/stdclass.cpp:173
        lock = <optimized out>
        rc = <optimized out>
#13 0x00007ff6e101b7b8 in PvrMessageQueue::dequeue (timeoutMs=<optimized out>, this=<optimized out>) at G:/msys64/home/B-S/flycast/core/hw/pvr/Renderer_if.cpp:149
        msg = <optimized out>
        msg = <optimized out>
        lock = <optimized out>
#14 PvrMessageQueue::waitAndExecute (timeoutMs=<optimized out>, this=<optimized out>) at G:/msys64/home/B-S/flycast/core/hw/pvr/Renderer_if.cpp:107
No locals.
#15 rend_single_frame (enabled=enabled@entry=@0x5fefdf: true) at G:/msys64/home/B-S/flycast/core/hw/pvr/Renderer_if.cpp:262
        timeout = 20
#16 0x00007ff6e0ee644f in Emulator::render (this=0x7ff6e1d69040 <emu>) at G:/msys64/home/B-S/flycast/core/emulator.cpp:989
No locals.
#17 0x00007ff6e111cadb in mainui_rend_frame () at G:/msys64/home/B-S/flycast/core/ui/mainui.cpp:57
No locals.
#18 0x00007ff6e111cde9 in mainui_loop () at G:/msys64/home/B-S/flycast/core/ui/mainui.cpp:95
        _ = <optimized out>
        currentRenderer = RenderType::DirectX11
#19 0x00007ff6e11e8e1e in SDL_main (argc=<optimized out>, argv=<optimized out>) at G:/msys64/home/B-S/flycast/core/windows/winmain.cpp:360
        _ = {args_ = std::vector of length 2, capacity 2 = {0x6beed0 "G:\\msys64\\home\\B-S\\flycast\\sa_build\\flycast.exe", 0x0}, arg_values_ = std::vector of length 1, capacity 1 = {{static buffer_size = 16, buffer_ = "\000ð­º\rð­º\rð­º\rð­º", data_ = 0x6beed0 "G:\\msys64\\home\\B-S\\flycast\\sa_build\\flycast.exe"}}, env_ = {static buffer_size = 256, buffer_ = "\000\000\000P\000\000\000\000\002\000\004\006", '\000' <repeats 20 times>, "0Ðx\000\000\000\000\000\f\000\000\000\000\000\000\000\005\000\000\000\000\000\000\000\f\000\000\000\000\000\000\000\f\000\000\000\000\000\000\000\000\000x\000\000\000\000\000\tø_\000\000\000\000\000©â0½ú\177\000\000è\002x\000\000\000\000\000Y\000S\000T\000E\000\001\0003\0002\000\\\000\f\000\000\000\000\000\000\000 \000\000\000\000\000\000\000_í1½ú\177\000\000\220Ðx\000\000\000\000\000_í1½ú\177\000\000\000\000x\000\000\000\000\000\tø_\000\000\000\000\000\200Ðx\000\000\000\000\000P\001x", '\000' <repeats 13 times>..., data_ = 0x0}, envp_ = std::vector of length 0, capacity 0, old_argc_ = 1, old_argv_ = 0x78dd60, old_env_ = 0x0, old_argc_ptr_ = 0x5ffe30, old_argv_ptr_ = 0x5ffe38, old_env_ptr_ = 0x0}
        tempDir = L"G:\\msys64\\tmp\\", '\000' <repeats 246 times>
        custom_entries = {{static kNameMaxLength = 64, static kValueMaxLength = 64, name = L"prod", '\000' <repeats 59 times>, value = L"Flycast", '\000' <repeats 56 times>}, {static kNameMaxLength = 64, static kValueMaxLength = 64, name = L"ver", '\000' <repeats 60 times>, value = L"v2.3.2-181-g308d9fc1a", '\000' <repeats 42 times>}}
        custom_info = {entries = 0x7ff6e36281c0 <SDL_main::custom_entries>, count = 2}
        handler = {filter_ = 0x0, callback_ = 0x7ff6e11e7d70 <dumpCallback(wchar_t const*, wchar_t const*, void*, EXCEPTION_POINTERS*, MDRawAssertionInfo*, bool)>, callback_context_ = 0x0, crash_generation_client_ = {ptr_ = 0x0}, dump_path_ = L"G:\\msys64\\tmp\\", next_minidump_id_ = L"242a9028-3666-40ec-b051-ea55eb3174cb", next_minidump_path_ = L"G:\\msys64\\tmp\\\\242a9028-3666-40ec-b051-ea55eb3174cb.dmp", dump_path_c_ = 0x2637fd0 L"G:\\msys64\\tmp\\", next_minidump_id_c_ = 0x2638020 L"242a9028-3666-40ec-b051-ea55eb3174cb", next_minidump_path_c_ = 0x26380a0 L"G:\\msys64\\tmp\\\\242a9028-3666-40ec-b051-ea55eb3174cb.dmp", dbghelp_module_ = 0x7ffaaa180000, minidump_write_dump_ = 0x7ffa972e6c40 <dbgcore!MiniDumpWriteDump>, dump_type_ = MiniDumpNormal, rpcrt4_module_ = 0x7ffabb7f0000, uuid_create_ = 0x7ffabb84b790 <UuidCreate>, handler_types_ = 7, previous_filter_ = 0x7ff6e1635770 <_gnu_exception_handler>, previous_pch_ = 0x0, handler_thread_ = 0x238, is_shutdown_ = false, handler_critical_section_ = {DebugInfo = 0xffffffffffffffff, LockCount = -1, RecursionCount = 0, OwningThread = 0x0, LockSemaphore = 0x0, SpinCount = 33556432}, handler_start_semaphore_ = 0x230, handler_finish_semaphore_ = 0x234, requesting_thread_id_ = 0, exception_info_ = 0x0, assertion_ = 0x0, handler_return_value_ = false, handle_debug_exceptions_ = true, consume_invalid_handle_exceptions_ = false, app_memory_info_ = std::__cxx11::list = {[0] = {ptr = 0, length = 0}}}
        __FUNCTION__ = "SDL_main"
        nws = {static buffer_size = 256, buffer_ = "G:\\msys64\\tmp\\\000\0004\000\\\000h\000o\000m\000e\000\\\000B\000-\000S\000\\\000f\000l\000y\000c\000a\000s\000t\000\\\000s\000a\000_\000b\000u\000i\000l\000d\000\\\000f\000l\000y\000c\000a\000s\000t\000.\000e\000x\000e", '\000' <repeats 11 times>, "F¥>½ú\177\000\000\000\000x\000\000\000\000\000c\000\000P", '\000' <repeats 20 times>, "ððx", '\000' <repeats 29 times>, "ððx\000\000\000\000\000\000\000x\000\000\000\000\000ððx\000\000\000\000\000"..., data_ = 0x5ff9e0 "G:\\msys64\\tmp\\"}
        tempDir8 = "G:\\msys64\\tmp\\"
        async = {<std::__basic_future<void>> = {<std::__future_base> = {<No data fields>}, _M_state = std::shared_ptr<std::__future_base::_State_baseV2> (use count 1, weak count 0) = {get() = 0x2719700}}, <No data fields>}
#20 0x00007ff6e125ff2e in main_getcmdline ()
No symbol table info available.
#21 0x00007ff6e0ee12ee in __tmainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:267
        lock_free = <optimized out>
        fiberid = <optimized out>
        nested = <optimized out>
#22 0x00007ff6e0ee13e6 in WinMainCRTStartup () at C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:157
        ret = 255
Continuing.
warning: 01:56:040 hw/holly/sb.cpp:220 N[SH4]: SB/HOLLY: System reset requested
```

**Comment by flyinghead** [comment]
The game or BIOS triggers a reboot of the system by writing to the SB_SFRES register. The remaining question is why? and debugging flycast probably won't help.
I was able to reproduce the reboot but even with full logging I couldn't spot anything suspicious before it happens.

**Comment by kihato** [comment]
Thanks for debugging and testing.

[Legacy of Kain - Soul Reaver v1.000 (1999)(EIDOS)(NTSC)(US)[!].state.zip](https://github.com/user-attachments/files/16745444/Legacy.of.Kain.-.Soul.Reaver.v1.000.1999.EIDOS.NTSC.US.state.zip)

When I tested while checking the commit graph, I found that there was a tree that did not reboot at all.
This is the tree that includes 57c88e0.
This was eventually merged in c146a92, so I was able to back test and identify the culprit, a7702c6.

This commit had many positive effects, but it may have been the only one that had a negative effect.
I did not expect it to affect the movie in the same way as #1272.
